src/operator/numpy/np_pad_op-inl.h

```diff
@@ -0,0 +1,735 @@
+  double constant_value;
+  DMLC_DECLARE_PARAMETER(NumpyPadParam) {
+    DMLC_DECLARE_FIELD(pad_width)
+        .describe("Number of values padded to the edges of each axis. "
+                  "((before_1, after_1), … (before_N,"
+                  "after_N)) unique pad widths for each axis. ((before, after),) "
+                  "yields same before and"
+                  "after pad for each axis. "
+                  "(pad,) or int is a shortcut for before = after = pad width for all"
+                  "axes.");
+    DMLC_DECLARE_FIELD(mode)
+        .set_default(1)
+        .describe("str or function, optional");
+    DMLC_DECLARE_FIELD(reflect_type)
+        .set_default("even")
+        .describe("Used in ‘reflect’, and ‘symmetric’. "
+                  "The ‘even’ style is the default with an unaltered reflection around "
+                  "the edge value. For the ‘odd’ style,"
+                  "the extended part of the array is created by subtracting the "
+                  "reflected values from two times the edge value.");
+    DMLC_DECLARE_FIELD(constant_value)
+        .set_default(0.0)
+        .describe("Used in ‘constant’. The values to set the padded values for each axis."
+                  "((before_1, after_1), ... (before_N, after_N)) unique pad constants for"
+                  "each axis."
+                  "((before, after),) yields same before and after constants for each axis."
+                  "(constant,) or constant is a shortcut for before = after = constant for all"
+                  "axes."
+                  "Default is 0.");
+  }
+};
+
+inline mxnet::TShape NumpyPadShapeImpl(const mxnet::TShape& ishape,
+                                       const mxnet::Tuple<Tuple<int>> pad_width) {
+  if (ishape.ndim() == 1) {
+    auto s = ishape[0] + pad_width[0][0] + pad_width[1][0];
+    return mxnet::TShape({s});
+  } else if (ishape.ndim() >= 2) {
+    int i;
+    int sshape_number = ishape.ndim();
+    mxnet::TShape oshape(ishape.ndim(), -1);
+    for (i = ishape.ndim() - 1; i >=0; i--) {
+      int base = ishape[i];
+      base = base + pad_width[i][0] + pad_width[i][1];
+      oshape[i] = base;
+    }
+  return oshape;
+  }
+  return mxnet::TShape({-1, -1});
+}
+
+inline bool NumpyPadOpShape(const nnvm::NodeAttrs& attrs,
+                            mxnet::ShapeVector* in_attrs,
+                            mxnet::ShapeVector* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 1U);
+  CHECK_EQ(out_attrs->size(), 1U);
+
+  const mxnet::TShape& ishape = (*in_attrs)[0];
+  if (!mxnet::ndim_is_known(ishape)) {
+    return false;
+  }
+  const NumpyPadParam& param = nnvm::get<NumpyPadParam>(attrs.parsed);
+
+  mxnet::TShape oshape = NumpyPadShapeImpl(ishape, param.pad_width);
+
+  if (shape_is_none(oshape)) {
+    LOG(FATAL) << "Pad does not exist.";
+  }
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, oshape);
+
+  return shape_is_known(out_attrs->at(0));
+}
+
+
+inline bool NumpyPadOpType(const nnvm::NodeAttrs &attrs,
```

Comment:
Move this function to `.cc`

Reply:
you can probably use `ElemwiseType<1, 1>` for this op instead of implementing your own